net-p2p/bitcoin-core/bitcoin-core-25.0.ebuild

```diff
@@ -0,0 +1,340 @@
+			share/examples/bitcoin.conf >"${ED}/etc/bitcoin/bitcoin.conf" || die
+		fowners bitcoin:bitcoin /etc/bitcoin/bitcoin.conf
+		fperms 0660 /etc/bitcoin/bitcoin.conf
+
+		newconfd contrib/init/bitcoind.openrcconf bitcoind
+		newinitd "${FILESDIR}/bitcoind.openrc" bitcoind
+		systemd_newunit contrib/init/bitcoind.service bitcoind.service
+
+		keepdir /var/lib/bitcoind
+		fperms 0750 /var/lib/bitcoind
+		fowners bitcoin:bitcoin /var/lib/bitcoind
+		dosym -r {/etc/bitcoin,/var/lib/bitcoind}/bitcoin.conf
+
+		insinto /etc/logrotate.d
+		newins "${FILESDIR}/bitcoind.logrotate-r1" bitcoind
+	fi
+
+	if use qt5 ; then
+		insinto /usr/share/icons/hicolor/scalable/apps
+		newins src/qt/res/src/bitcoin.svg bitcoin128.svg
+
+		domenu "${FILESDIR}/org.bitcoin.bitcoin-qt.desktop"
+
+		if use kde ; then
+			insinto /usr/share/kservices5
+			doins "${FILESDIR}/bitcoin-qt.protocol"
+		fi
+	fi
+
+	if use examples ; then
+		docinto examples
+		dodoc -r contrib/{linearize,qos}
+		use zeromq && dodoc -r contrib/zmq
+	fi
+}
+
+efmt() {
+	: ${1:?} ; local l ; while read -r l ; do "${!#}" "${l}" ; done < <(fmt "${@:1:$#-1}")
+}
+
+pkg_preinst() {
+	if use daemon && [[ -d "${EROOT}/var/lib/bitcoin/.bitcoin" ]] ; then
+		if [[ -h "${EROOT}/var/lib/bitcoin/.bitcoin" ]] ; then
+			dosym -r /var/lib/bitcoin{d,/.bitcoin}
+		elif [[ ! -e "${EROOT}/var/lib/bitcoind" || -h "${EROOT}/var/lib/bitcoind" ]] ; then
+			efmt ewarn <<-EOF
+				Your bitcoind data directory is located at ${EPREFIX}/var/lib/bitcoin/.bitcoin,
+				a deprecated location. To perform an automated migration to
+				${EPREFIX}/var/lib/bitcoind, first shut down any running bitcoind instances
+				that may be using the deprecated path, and then run:
+
+				# emerge --config ${CATEGORY}/${PN}
+				EOF
+			insinto /var/lib/bitcoin
+			mv -- "${ED}/var/lib/bitcoin"{d,/.bitcoin} || die
+			dosym -r {/etc/,/var/lib/bitcoin/.}bitcoin/bitcoin.conf
+			dosym -r /var/lib/bitcoin{/.bitcoin,d}
+		fi
+	fi
+
+	if use kde && [[ -d "${EROOT}/usr/share/kde4" ]] ; then
+		dosym -r /usr/share/{kservices5,kde4/services}/bitcoin-qt.protocol
+	fi
+}
+
+pkg_postinst() {
+	# we don't use xdg.eclass because it adds unconditional IDEPENDs
+	if use qt5 ; then
+		xdg_desktop_database_update
+		xdg_icon_cache_update
+	fi
+
+	if use daemon && [[ -z "${REPLACING_VERSIONS}" ]] ; then
+		efmt -su elog <<-EOF
+			To have ${PN} automatically use Tor when it's running, be sure your \
+			'torrc' config file has 'ControlPort' and 'CookieAuthentication' set up \
+			correctly, and:
+			- Using an init script: add the 'bitcoin' user to the 'tor' user group.
+			- Running bitcoind directly: add that user to the 'tor' user group.
+			EOF
+	fi
+}
+
+pkg_postrm() {
+	if use qt5 ; then
+		xdg_desktop_database_update
+		xdg_icon_cache_update
+	fi
+}
+
+pkg_config() {
+	if [[ -d "${EROOT}/var/lib/bitcoin/.bitcoin" && ! -h "${EROOT}/var/lib/bitcoin/.bitcoin" ]] &&
+		[[ ! -e "${EROOT}/var/lib/bitcoind" || -h "${EROOT}/var/lib/bitcoind" ]]
```

Comment:
I am sorry if this was discussed before, but why is the `-h "${EROOT}/var/lib/bitcoind"` condition necessary? Why should the new home exists and be a symlink? And why would we then still migrate and not simply always bail out if `/var/lib/bitcoind` already exists?

Reply:
`pkg_preinst()` creates a symlink at `${EROOT}/var/lib/bitcoind`, pointing (relatively) at `${EROOT}/var/lib/bitcoin/.bitcoin`, so that the system service unit/runscript will continue to work even before the migration has been performed. (Indeed, the user could opt _never_ to perform the migration.) Without the `-h "${EROOT}/var/lib/bitcoind"` condition here, we would always fail to perform the migration as long as the compatibility symlink (which we installed!) was remaining in place.